---

title: "Acidovorax Pangenome study"
author: "Siani Roberto"
date: "28.01.2021"
output: 
  html_document:
    theme: cosmo
    highlight: monochrome
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

### SETUP

```{r, include = F}

## load required libraries

if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse,
               ggpubr,
               FactoMineR,
               factoextra,
               caret,
               openxlsx,
               EnhancedVolcano,
               doParallel,
               patchwork,
               hrbrthemes,
               superheat)

## choose palette for visualization

my_palette = get_palette(c("#50b47b", # green for commensal
                           "#7e60b7", # purple for free-living
                           "#c06846"), # red for pathogen 
                         3)

## coerce variables to binary distribution

binaryTransform = function(x){
  if (x >= 1) {return(1)}
  else{return(x)}
} 

## check if all the variables are normally distributed

checkNormality = function(x){
  all(lapply(x, shapiro.test) %>% 
        sapply(`[`, "p.value") < 0.05)
}

## list of the groups for stat_compare_means()

my_comparison = list(c("commensal", "free-living"),
                     c("free-living", "pathogen"),
                     c("pathogen", "commensal"))

```

### BIOASSAYS

```{r}

## load results from L.japonicus bioassays

IN = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx",
                sheet = "growth_assay",
                sep.names = " ") %>%
  select(c(bacteria, 
           `plant part`, 
           `Length`, 
           `Fresh Weight`, 
           `Dry Weight`, 
           `Total Fresh Weight`, 
           `Total Dry Weight`,
           `Total Length`)) %>% 
  mutate(coloring = ifelse(bacteria %in% "no bacteria", "Control", "Bacteria"))

## convert to tidy frame

DF = IN %>% 
  pivot_longer(cols = 3:8,
               names_to = "metric") %>% 
  na.omit() %>% 
  group_by(bacteria, metric, `plant part`)

## fix categorical variables values 

DF$`plant part`[DF$metric %in% "Total Length" | DF$metric %in% "Total Fresh Weight" | DF$metric %in% "Total Dry Weight"] = ""

rm(IN)

```

```{r}

## explore all the metrics for significance

p1a = ggplot(DF[DF$metric %in% "Total Fresh Weight", ],
             aes(x = reorder(bacteria, value, FUN = median),
                   y = value,
                   fill = coloring,
                   alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.65)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Total Fresh Weight" &
                                      DF$coloring %in% "Control", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Total Fresh Weight (g)")

p1a + 
  plot_layout()

p1b = ggplot(DF[DF$metric %in% "Total Length", ],
             aes(x = reorder(bacteria, value, FUN = median),
                   y = value,
                   fill = coloring,
                   alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(100, 350)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Total Length" &
                                      DF$coloring %in% "Control", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Total Length (mm)")

p1b + 
  plot_layout()

p1c = ggplot(DF[DF$metric %in% "Length" &
                  DF$`plant part` %in% "Shoot", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 150)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Length" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Shoot", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Shoot Length (mm)")

p1d = ggplot(DF[DF$metric %in% "Length" &
                  DF$`plant part` %in% "Root", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 250)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Length" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Root", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Root Length (mm)")

p1c / p1d

p1e = ggplot(DF[DF$metric %in% "Fresh Weight" &
                  DF$`plant part` %in% "Shoot", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.3)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Fresh Weight" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Shoot", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Shoot Fresh Weight (g)")

p1f = ggplot(DF[DF$metric %in% "Fresh Weight" &
                  DF$`plant part` %in% "Root", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.4)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Fresh Weight" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Root", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Root Fresh Weight (g)")

p1e / p1f

p1g = ggplot(DF[DF$metric %in% "Total Dry Weight", ],
             aes(x = reorder(bacteria, value, FUN = median),
                   y = value,
                   fill = coloring,
                   alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.100)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Total Dry Weight" &
                                      DF$coloring %in% "Control", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Total Dry Weight (g)")

p1g + 
  plot_layout()

p1h = ggplot(DF[DF$metric %in% "Dry Weight" &
                  DF$`plant part` %in% "Shoot", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.1)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Dry Weight" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Shoot", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Shoot Dry Weight (g)")

p1i = ggplot(DF[DF$metric %in% "Dry Weight" &
                  DF$`plant part` %in% "Root", ],
             aes(x = reorder(bacteria, value, FUN = median),
                 y = value,
                 fill = coloring,
                 alpha = 0.9)) +
  geom_boxplot(outlier.shape = NA) +
  theme_ipsum_rc(plot_margin = margin(1, 1, 1, 1),
                 grid = "XY") +
  ylim(c(0, 0.05)) +
  scale_fill_manual(values = c("#008080",
                               "#800000")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 0.5,
                                   vjust = 0.75,
                                   size = 10,
                                   face = "bold"),
        axis.title.x = element_blank()) +
  stat_compare_means(ref.group = "no bacteria",
                     label = "p.signif",
                     hide.ns = T,
                     method.args = list(p.adjust.methods = "holm"),
                     method = "t.test") +
  geom_hline(yintercept = median(DF[DF$metric %in% "Dry Weight" &
                                      DF$coloring %in% "Control"  &
                                      DF$`plant part` %in% "Root", ]$value),
             color = "#800000",
             linetype = "dotted",
             size = 1.5,
             alpha = 0.6) +
  ylab("Root Dry Weight (g)")

p1h / p1i

rm(p1a,
   p1b,
   p1c,
   p1d,
   p1e,
   p1f)

```

```{r}

## final plot for the bioassays

DF_WIDE = DF %>% 
  summarise(mean = mean(value)) %>% 
  pivot_wider(names_from = c(`plant part`, metric),
              values_from = mean,
              names_sep = " ")

tDF = DF_WIDE[, 2:10] %>% 
  t() %>% 
  data.frame

colnames(tDF) = DF_WIDE$bacteria

## load summarized results from the tests

effects = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx",
                sheet = "growth_effects",
                sep.names = " ") %>% 
  select(bacteria, effect)

tDF = tDF %>% 
  apply(MARGIN = 2,
        FUN = function(x) 
          (((x - tDF$`no bacteria`) 
            / tDF$`no bacteria`) * 100)) %>% 
  t %>% 
  as.data.frame() %>% 
  mutate(bacteria = DF_WIDE$bacteria) %>% 
  filter(!bacteria %in% "no bacteria") %>%
  merge(effects,
        by.y = "bacteria")

rownames(tDF) = tDF$bacteria

## plot heatmap

superheat(tDF[, 2:10],
          scale = T,
          heat.pal = get_palette("npg", 3),
          membership.rows = tDF$effect,
          pretty.order.rows = T,
          smooth.heat = T,
          left.label.text.alignment = "left",
          left.label.col = "white",
          bottom.label.col = "white",
          grid.hline.col = "white",
          grid.vline.col = "white",
          padding = 0.1,
          legend.text.size = 0,
          left.label = "variable",
          grid.vline.size = 1,
          bottom.label.size = 0.05,
          left.label.size = 0.05)

rm(DF_WIDE,
   tDF,
   effects,
   DF)

```

```{r}

## load panMatrix for L.japonicus isolates only

IN = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx", 
                sheet = "panMatrix_LJR",
                colNames = T,
                rowNames = T) %>% 
  t

## conver to tidy frame
  
DF = data.frame(IN, 
                row.names = c(rownames(IN))) %>% 
  mutate(effect = ifelse(rownames(.) %in% c("LjRoot120", "LjRoot124", "LjRoot176", "LjRoot75", "LjRoot87", "LjRoot88", "LjRoot103"), "robust", "neutral")) %>% 
  t %>% 
  as.data.frame()

## write table to submit to Galaxy LEfSE

write.table(DF,
          file = "forLefsa.tsv",
          sep = "\t",
          col.names = F,
          quote = F)

## load summirized results

res.lda = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx", 
                sheet = "res.lefse",
                colNames = T,
                rowNames = F,
                sep.names = " ") %>% 
  group_by(Outcome)


## plot results

ggbarplot(res.lda, 
          x = "Pfam", 
          y = "Logarithmic LDA score",
          fill = "Family",
          palette = get_palette("npg", 4),
          sort.val = "asc",
          alpha = 0.66,
          ylab = "Logarithmic LDA score",
          lab.col = "associated_group") +
  theme_ipsum(grid = "XY",
              plot_margin = margin(0, 0, 0, 0),
              base_size = 20,
              axis_title_size = 20) +
  theme(legend.title = element_blank(),
        axis.title.y = element_blank()) +
  coord_flip() +
  plot_layout()

```

### GENOME ARCHITECTURE

```{r}

## load summarized info for the genomes

DF0 = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx",
                sheet = "summary",
                rowNames = F,
                colNames = T,
                sep.names = " ") %>% 
  mutate(logTotalLength = log10(total_length)) %>% 
  group_by(phenotype)

## plot a dot-chart of the genome length

A = ggdotchart(DF0,
                 x = "layers",
                 y = "logTotalLength",
                 color = "phenotype",
                 palette = my_palette,
                 group = "phenotype",
                 remove = "LjRoot141",
                 x.text.col = T,
                 xlab = "",
                 ylab = "Logarithmic Genome Lenght",
                 ggtheme = theme_ipsum_rc("X",
                                          plot_margin = margin(0, 0, 0, 0),
                                          axis_title_size = 15,
                                          base_size = 15)) +
  theme(legend.position = "none")

## plot a box-plot of genome length by group

B = ggboxplot(DF0,
                "phenotype",
                "logTotalLength",
                group = "phenotype",
                fill = "phenotype",
                palette = my_palette[c(2, 1, 3)],
                order = c("free-living", "commensal", "pathogen"),
                xlab = "",
                ylab = "Logarithmic Genome Lenght",
                ggtheme = theme_ipsum_rc(plot_margin = margin(0, 0, 0, 0),
                 axis_title_size = 15,
                 base_size = 20)) +
  stat_compare_means(comparisons = my_comparison,
                     label = "p.signif") +
  theme(legend.position = "none")

## plot a box-plot of number of genes per kb by group

C = ggboxplot(DF0,
                "phenotype",
                "num_genes_per_kb",
                group = "phenotype",
                fill = "phenotype",
                palette = my_palette[c(2, 1, 3)],
                order = c("free-living", "commensal", "pathogen"),
                xlab = "",
                ylab = "Number of genes per kb",
                ggtheme = theme_ipsum_rc(plot_margin = margin(0, 0, 0, 0),
                 axis_title_size = 15,
                 base_size = 20)) +
  stat_compare_means(comparisons = my_comparison,
                     label = "p.signif") +
  theme(legend.position = "none")

## assemble plots

A /
  (B | C) +
  plot_layout(height = c(1, 4)) +
  plot_annotation(tag_levels = "a",
                  theme = theme_ipsum_rc())


## extract the observed phenotype and source for later

phenotype = DF0 %>% 
  dplyr::select(layers, phenotype) %>% 
  as.data.frame() %>% 
  mutate(phenotype = as.factor(phenotype),
         ID = layers) %>% 
  select(!layers)

source = DF0 %>% 
  dplyr::select(layers, source) %>% 
  as.data.frame() %>% 
  mutate(source = as.factor(source),
         ID = layers) %>% 
  select(!layers)


## clean environment

rm(DF0,
   A,
   B,
   C)

```


### PRE-PROCESSING

```{r}

## source data as a table, transpose and convert in binary matrix (presence/absence)

TAB = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx", 
                sheet = "panMatrix",
                colNames = T,
                rowNames = T,
                sep.names = " ") %>% 
  t() %>%
  as.matrix() %>%
  apply(MARGIN = c(1,2), 
        FUN = function(x) binaryTransform(x))


## preprocess by removing near-zero and zero variance column (I am not sure if corr is working ...)

DF = TAB %>% 
  preProcess(method = c("nzv", 
                        "zv", 
                        "corr"),
             verbose = T) %>% 
  predict(TAB) %>% 
  as.data.frame()

rownames(DF) = phenotype$ID

## check normality for every variable (TRUE if all the variable are normal. Else FALSE: adjust preProcess arguments)

checkNormality(DF)

## clean-up environment

rm(TAB)

```

### PRINCIPAL COMPONENT ANALYSIS

```{r}

## PCA from FactoMineR, keeping 3 dimensions

res.pca = PCA(DF, 
          scale.unit = F,
          graph = F,
          ncp = 3)

## extract results for individuals

ind = get_pca_ind(res.pca)$coord %>% 
  as.data.frame() %>% 
  mutate(ID = rownames(.)) %>% 
  merge(source, by =  "ID")

checkNormality(ind %>% 
                 select(c(Dim.1, Dim.2, Dim.3)))

## test significance of distance between clusters

res.adonis = vegan::adonis(data = ind %>% 
                             select(phenotype),
                           formula = ind %>% 
                             select(Dim.1, 
                                    Dim.2, 
                                    Dim.3) ~ phenotype,
                           method = "eu",
                           permutations = 1000)

## ordination plot

A = ind %>% 
  ggplot(aes(x = Dim.1,
             y = Dim.2,
             group = phenotype,
             color = phenotype,
             label = ID,
             shape = phenotype)) +
  geom_point(size = 3) +
  stat_ellipse(linetype = "dotted",
               geom = "polygon",
               type = "norm",
               alpha = 0.1,
               aes(fill = phenotype)) +
  theme_ipsum_rc(plot_margin = margin(0, 0, 0, 0),
                 base_size = 20,
                 axis_title_size = 20) +
  scale_color_manual(values = my_palette) +
  scale_fill_manual(values = my_palette) +
  theme(legend.position = "none") +
  labs(x = "Dim.1 (24.3%)",
       y = "Dim.2 (12.4%)")

B = ind %>% 
  ggplot(aes(x = Dim.1,
             y = Dim.3,
             group = phenotype,
             color = phenotype,
             label = ID, 
             shape = phenotype)) +
  geom_point(size = 3) +
  stat_ellipse(linetype = "dotted",
               geom = "polygon",
               type = "norm",
               alpha = 0.1,
               aes(fill = phenotype)) +
  theme_ipsum_rc(plot_margin = margin(0, 0, 0, 0),
                 base_size = 20,
                 axis_title_size = 20) +
  scale_color_manual(values = my_palette) +
  scale_fill_manual(values = my_palette) +
  labs(x = "Dim.1 (24.3%)",
       y = "Dim.3 (10.3%)") +
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 20))

## get contributions for the variables

cntrb = rbind(get_pca_var(res.pca)$contrib %>% 
                as.data.frame() %>%
                slice_max(Dim.1), 
              get_pca_var(res.pca)$contrib %>% 
                as.data.frame() %>%
                slice_max(Dim.2), 
              get_pca_var(res.pca)$contrib %>% 
                as.data.frame() %>%
                slice_max(Dim.3)) %>% 
  mutate(Pfam = rownames(.)) %>% 
  pivot_longer(!Pfam, 
               names_to = "Dimension",
               values_to = "Contribution") %>% 
  arrange(Dimension)

C = ggbarplot(cntrb,
                x = "Pfam",
                y = "Contribution",
                fill = "Dimension",
                position = position_dodge(),
                width = 0.90,
                alpha = 0.66,
                palette = get_palette("Blues", 3)) +
  theme_ipsum_rc(plot_margin = margin(0, 0, 0, 0),
                 base_size = 15,
                 axis_title_size = 20) +
  theme(panel.grid.major.x = element_blank(),
        axis.text.x = element_text(angle = 60,
                                   hjust = 1,
                                   vjust = 1),
        axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 20))

(A | B) /
  C + plot_layout(heights = c(4, 1)) +
  plot_annotation(tag_levels = "a",
                  theme = theme_ipsum_rc())

## cleaning environment

rm(cntrb,
   ind,
   res.pca,
   A,
   B,
   C)

```

### GENE CLUSTERS FUNCTIONS ENRICHMENT ANALYSIS

```{r}

## load enriched functions data

DF3 = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx",
                sheet = "enrichedFunctions",
                sep.names = " ") %>% 
  filter(!associated_groups %in% "free-living",
         !associated_groups %in% "pathogen, free-living",
         !associated_groups %in% "commensal, free-living",
         !associated_groups %in% "commensal, pathogen") %>% 
  na.exclude() %>% 
  mutate(log2fold = log2(enrichment_score))

## reverse enrichment score for one of the groups

DF3$log2fold[DF3$associated_groups %in% "commensal"] = 0 - DF3$log2fold[DF3$associated_groups %in% "commensal"]


## load enriched functions of interest data

DF4 = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx",
                sheet = "curatedList") %>% 
  mutate(log2fold = log2(enrichment_score))

## reverse enrichment score for one of the groups
 
DF4$log2fold[DF4$associated_groups %in% "commensal"] = 0 - DF4$log2fold[DF4$associated_groups %in% "commensal"]

## volcano plot with enrichment score x adjusted q value

A = EnhancedVolcano(DF3,
                    lab = DF3$Pfam,
                    selectLab = DF4$Pfam,
                    x = "log2fold",
                    y = "adjusted_q_value",
                    cutoffLineType = "dotted",
                    pointSize = 3,
                    FCcutoff = NA,
                    ylim = c(0, 20),
                    xlim = c(-10, 15),
                    colAlpha = 0.33,
                    labSize = 0,
                    pCutoff = 0.05,
                    col = c("#2a5b6f",
                            "#000000",
                            "#cc6092")) +
  labs(y = bquote(~-Log[10]~ "q-value")) +
  theme_ipsum_rc(grid = "XY",
                 plot_margin = margin(0, 0, 0, 0),
                 axis_title_size = 20) +
  theme(legend.position = "none",
        plot.title = element_blank(),
        plot.subtitle = element_blank())

## dot-chart of enrichment score for all the functions of interest

B = ggdotchart(DF4,
               x = "Pfam",
               y = "log2fold",
               color = "Family",
               sorting = "asc",
               palette = get_palette("npg", 4),
               rotate = F,
               dot.size = 2.5,
               y.text.col = T,
               ggtheme = theme_ipsum_rc(grid = "X",
                                        plot_margin = margin(0, 0, 0, 0),
                                        base_size = 20,
                                        axis_title_size = 20),
               font.tickslab = c(4),
               ticks = F,
               ylab = "Enrichment Score") +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(size = 12,
                                   angle = 80,
                                   vjust = 1),
        axis.title.x = element_blank())

## combined graph

A + B +
  plot_layout(heights = c(4, 1)) +
  plot_annotation(tag_levels = "a",
                  theme = theme_ipsum_rc())

## clean environment

rm(DF3, 
   DF4,
   A,
   B)

```

### BIOSYNTHESIS GENE CLUSTERS

```{r}

## import antiSMASH results

DF2 = read.xlsx("~/Desktop/AVX_PGC/IN/res_AVX.xlsx", 
                        sheet = "antiSMASH",
                sep.names = " ") %>% 
  merge(phenotype, by = "ID") %>% 
  select(!ID) %>% 
  pivot_wider(names_from = BGC,
              values_from = count,
              values_fill = 0,
              values_fn = {sum}) %>% 
  pivot_longer(!phenotype,
               values_to = "value",
               names_to = "BGC") %>% 
  arrange(BGC)

## transform to percentages in this shitty way cause I don't know any better

DF2$value[DF2$phenotype == "free-living"] = 
  round(DF2$value[DF2$phenotype == "free-living"] / 23 * 100)
DF2$value[DF2$phenotype == "pathogen"] = 
  round(DF2$value[DF2$phenotype == "pathogen"] / 21 * 100)
DF2$value[DF2$phenotype == "commensal"] = 
  round(DF2$value[DF2$phenotype == "commensal"] / 62 * 100)

## visualize bar-plot of BGC frequency per cluster

ggbarplot(DF2,
          x = "BGC",
          y = "value",
          fill = "phenotype",
          position = position_dodge2(),
          palette = my_palette,
          label = T,
          ylab = "% of genomes",
          width = 0.85,
          alpha = 0.66,
          ggtheme = theme_ipsum_rc(grid = F,
                                   plot_margin = margin(0, 0, 0, 0),
                                   axis_title_size = 20,
                                   base_size = 20)) +
  theme(axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 30,
                                   hjust = 0.5,
                                   vjust = 0.95),
        legend.title = element_blank()) +
  plot_layout()

## clean environment

rm(DF2)

```

### MACHINE LEARNING

```{r}

## prepare dataframe for analysis

MLDF = DF %>%
  mutate(observation = factor(phenotype$phenotype))

levels(MLDF$observation)[levels(MLDF$observation) == "free-living"] = "freeLiving"

## create training and testing data partitions

set.seed(1)

inTrain = createDataPartition(MLDF$observation,
                              p = .9,
                              list = F)

trDF = MLDF[inTrain,]
teDF = MLDF[-inTrain,]

## set up multithreading

cl = makeCluster(6, 
                 type = "FORK")

registerDoParallel(cl)

```

### NEURAL NETWORKS WITH FEATURE EXTRACTION

```{r}

## prepare seed vector for reproducibility

set.seed(1)

seeds <- vector(mode = "list", length = 1000)

for (i in 1:1000) seeds[[i]] <- sample.int(1000, 39)

## For the last model:

seeds[[1001]] <- sample.int(1000, 1)

## train classifier on training partition

set.seed(1)

model_NN = train(observation ~ .,
                 data = trDF,
                 method = "pcaNNet",
                 trControl = trainControl(method = "LOOCV",
                                          number = 1000,
                                          seeds = seeds,
                                          classProbs = T,
                                          returnResamp = "final",
                                          summaryFunction = multiClassSummary,
                                          trim = F),
                 tuneLenght = 15,
                 metric = "Accuracy")

## The final values used for the model were sigma = 0.008374599 and C = 885.4354.

## sigma = 0.0002142411 and C = 0.25

model_NN

## test model on testing partition

predicted = predict(model_NN, 
                    teDF,
                    type = "prob")

votes = predict(model_NN,
                teDF)

confusionMatrix(reference = teDF$observation,
                data = votes,
                mode = "everything",
                positive = "pathogen")

## re-test model on full collection
 
predicted = predict(model_NN, 
                    MLDF,
                    type = "prob")

votes = predict(model_NN,
                MLDF)

confusionMatrix(reference = MLDF$observation,
                data = votes,
                mode = "everything",
                positive = "pathogen")

```

### SESSION INFO

```{r session-info, include=TRUE, echo=TRUE, results='markup'}

devtools::session_info()

```

